ompi/mca/coll/base/coll_base_util.c

```diff
@@ -558,3 +559,69 @@ const char* mca_coll_base_colltype_to_str(int collid)
     }
     return colltype_translation_table[collid];
 }
+
+OBJ_CLASS_INSTANCE(ompi_coll_base_hostname_item_t, opal_list_item_t, NULL, NULL);
+
+int
+ompi_coll_base_get_nnodes(struct ompi_communicator_t *comm)
+{
+    ompi_group_t* group;
+    ompi_proc_t *proc;
+    opal_list_t hostname_list;
+    ompi_coll_base_hostname_item_t *host_item, *next, *new_item;
+    bool already_in_the_list;
+    char* hostname;
+    int i, cmp, group_size, nodes_nb;
+
+    OPAL_OUTPUT((ompi_coll_base_framework.framework_output,"coll get_nnodes called."));
+    OBJ_CONSTRUCT(&hostname_list, opal_list_t);
+    /* Is inter communicator ? */
+    if (OMPI_COMM_IS_INTER(comm)) {
+        group = comm->c_remote_group;
+    } else {
+        group = comm->c_local_group;
+    }
+    /* allocate an array of node id */
+    group_size = ompi_group_size(group);
+    OPAL_OUTPUT((ompi_coll_base_framework.framework_output,"coll get_nnodes group size %d.", group_size));
+    /* For each rank */
+    for (i=0 ; i<group_size ; i++) {
+        proc = ompi_group_get_proc_ptr (group, i, true);
+        hostname = opal_get_proc_hostname(&proc->super);
```

Comment:
@rhc54 @jjhursey `ompi_coll_base_get_nnodes()` returns the number of nodes used by a given communicator.
I do not expect `PMIx` to be MPI communicators aware.

That being said, the current implementation calls `opal_get_proc_hostname()` for each rank of the communicator. If I remember correctly, we tried to avoid that (because of both performance and memory overhead). A more efficient implementation would be to use a bitmap, use the nodeid (e.g. an `int`) of each proc to set the corresponding bit, and then count the number of bits in the bitmap. Can `PMIx` be used to retrieve the nodeid of each proc? (bonus point if `PMIx` can return the nodeids of a list or proc).

Note both current and proposed implementation do not use any collective operations.

I am pretty sure `ROMIO` and possibly `ompio` implement a similar subroutine, and would benefit from a single and optimized implementation.

Reply:
> I do not expect PMIx to be MPI communicators aware.

Yeah, it definitely won't be _unless_ PMIx_Connect was called when those communicators are formed. We do that for inter-communicators, so PMIx does know that membership. I'm pretty sure we don't do it for intra-communicators since those procs are already "connected" by definition, so we wouldn't know about the membership there.

> Can PMIx be used to retrieve the nodeid of each proc?

Sure - just call `PMIx_Get` with the name of the proc and the `PMIX_NODEID` key

> (bonus point if PMIx can return the nodeids of a list or proc).

You can, but you would need to use a different interface. You would `PMIx_Query_info` with the `PMIX_NODEID` key and an array of proc names. I'd have to add the backend support for it.

Reply:
No need for all this fanciness, MPI provides support for doing all this in MPI. Similar capability is already available in HAN in the subcomm file.

Reply:
HAN uses `MPI_Allreduce()` and `MPI_Allgather()` to get this kind of information.

Wouldn't it be faster to get node ids from `PMIx` and then have each task independently process them to achieve the very same outcome?

Reply:
I would rather not rely on yet another component for this kind of information. The MPI version is portable, and will remain so whatever runtime we are using. Plus, HAN's code does more than just what is required here, you only need a comm_split and a allgather. Moreover, if we are taking this route and will almost always need the leader and local communicators, we can embed their creation into the communicator creation, bringing the cost to a minimum.

Reply:
I'm not sure I fully understand this concern. OMPI already relies on the RM to provide a wide range of info. If that RM provides info relevant to HAN or any other part of MPI, it seems odd to ignore it for the sake of purity. Whether it is faster to collect from the RM vs doing two collectives ops is a different question - I don't know the answer to it, but it might at least merit a quick check to see.

If the RM path proves to be faster, then perhaps an alternate solution would be to first see if the RM provided the info (a quick "get optional" on one nodeid would resolve that as the data is in shared memory, and if one nodeid is provided then so will all the others) and then do the collectives if it isn't there.

Reply:
I'm afraid you misinterpreted a large part of the discussion here, nobody (and certainly not HAN) is ignoring the RM information in order to reinvent the wheel, we already have it in a format that is OMPI-specific and that we can use in a more optimized way than linearly going through the PMIx database.

Reply:
Let's discuss this on Tuesday's webex -- I don't have the larger context here, but looking at `ompi_coll_base_get_nnodes()`, I'm not quite grokking how a comm_split+allgather+string compares to test for "unknown" hostnames (which could lead to errors) and further nested string lookup loops would be faster than the put-the-node-id-into-a-bit-field approach.

Reply:
We talked about this on the webex today, and things were much more clear to me:

1. To be clear: @bosilca doesn't think that this functionality belongs here in tuned, anyway (please correct me if I got that sentiment incorrect).
2. But if it *is* to be used here in `tuned`, we should probably extract the functionality out of HAN's init-time startup where this information (and some additional stuff) is already obtained, and put that in coll base.  That way, other coll components can use it.  From my understanding, the HAN implementation is already good/scalable/happy; it isn't an issue of PMIx vs. MPI -- it was more: we already have the information locally; there's no need to calculate that information again.
3. We all recognize that this PR probably needs to be re-filed on master.

Reply:
In my opinion topologic configuration should be avalaible in tuned. Tha main reason is that there is collective which does not have han implementation. Then, it's not guarantee that all collectives can be easily implemented without risk of bugs using han approach (alltoall with disordered ranks,etc.). In that cases fallback will be used and tuned implementations would be used and it's can be optimized using topologic configuration. In the worst case, that kind of configuration might be used in tuned fallback situations. This is quite true for libnbc.

Then, i'm not against the idea of merging topological init of Han and tuned. It implies that we consider that Han is the most of the time enabled and used and thus subcomms init is essential anyway. It's not a problem to me even if our customer does not use han right now (it may change soon). In such approach comm rules init in tuned must be delayed (same in libnbc of course). It implies a lot of change but it can facilitate the acceptability of this PR it's does not matter.

I will resubmit this PR on master branch anyway but before doing this i want to be sure that we agree about the global scope of what this PR can address. I'm at your disposal to find an agreement.

Reply:
I will evaluate Coll/han vs coll/tuned and try to illustrate on which case (at the moment) node level configuration on tuned can be useful. I imagine that there is an interest when tuned is used for both intra node level and inter node level with the same number of processes. In this situation the current file format can't differentiate both configuration resulting on non optimal performances. Then, depending on that results i will decide to abandon node level format or not.